### ISSUE TITLE
Complete collection of syntax highlighting test files

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/bat#1213]
*With all the recent news about Hacktoberfest I thought it would be a good idea to point out good beginner issues that would be actually helpful for `bat`. In the past years, I have actually experienced Hacktoberfest as a really great event - both as a contributor as well as a maintainer.*

As of [recently](https://github.com/sharkdp/bat/pull/1126), `bat` has a set of syntax highlighting regression tests (see #1124 for more details). The main idea is that we have a large collection of test files for each and every language that `bat` can highlight. This way, we can make sure that we do not run into issues we had in the past where either (1) syntax highlighting for some language is suddenly not working anymore or (2) `bat` suddenly crashes/panics for some input (due to incompatibilities in the `regex` flavors in `syntect` and Sublime Text).

In order to add a new test file, you can follow these steps (let's take "Ruby" as an example):

1. Make sure that you are running the **latest version of `bat`** (`master` or `bat` 0.16) and that `bat` is available on the path.
2. Find an example Ruby source file or write one yourself. If possible, the file should aim to be "comprehensive" (i.e. include a lot of the possible syntax), but this is not strictly necessary. A simple file is better than none at all. Also, the files shouldn't be gigantic.
3. Save the file in `tests/syntax-tests/source/Ruby` (adapt for your language). The file name could be `test.rb` (adapt extension) but can also be adapted if that is necessary in order for `bat` to highlight it correctly (e.g. `Makefile`).
4. If you have copied the file from somewhere else, please make sure that the file *may* be copied under the respective license and that the license is compatible with `bat`s license. If it requires attribution, please add a `LICENSE.md` in the same folder with a text like this:
    ```
    The `test.rb` file has been added from [enter source here] under the following license: 
    
    [add license text here]
    ```
5. Go to `tests/syntax-tests` and run the `update.sh` Bash script. A new file should be generated in the `highlighted` folder (e.g. `highlighted/Ruby/test.rb`).
6. Use `cat` or `bat --language=txt` to display the content of this file and make sure that the syntax highlighting looks correct.
7. `git add` the new files in the `source` folder as well as the autogenerated files in the `highlighted` folder.
8. Commit and submit a PR! Please reference this issue (#1213).

List of languages / syntaxes:
- [x] **ActionScript**
- [x] Apache Conf
- [x] **AppleScript**
- [x] ARM Assembly
- [x] AsciiDoc (Asciidoctor)
- [x] **ASP**
- [x] Assembly (x86_64)
- [x] AWK
- [x] **Batch File**
- [x] BibTeX
- [x] **Bourne Again Shell (bash)**
- [x] C
- [x] **C#**
- [x] **C++**
- [x] Cabal
- [x] **Clojure**
- [x] CMake
- [x] CoffeeScript
- [x] CpuInfo (`/proc/cpuinfo`)
- [x] Crystal
- [x] **CSS**
- [x] CSV
- [x] **D**
- [x] Dart
- [x] **Diff**
- [x] Dockerfile
- [x] DotENV
- [x] Elixir
- [x] Elm
- [x] Email
- [x] **Erlang**
- [x] F#
- [x] Fortran
- [x] Friendly Interactive Shell (fish)
- [x] fstab
- [x] Git Attributes
- [x] Git Config
- [x] Git Ignore
- [x] GLSL
- [x] **Go**
- [x] GraphQL
- [x] **Graphviz (DOT)**
- [x] **Groovy**
- [x] `/etc/group`
- [x] **Haskell**
- [x] Highlight non-printables (`--show-all`)
- [x] `/etc/hosts`
- [x] **HTML**
- [x] INI
- [x] **Java**
- [x] Java Server Page (JSP)
- [x] **JavaScript**
- [x] Jinja2
- [x] **JSON**
- [x] jsonnet
- [x] Julia
- [x] Kotlin
- [x] **LaTeX**
- [x] Less
- [x] **Lisp**
- [x] Literate Haskell
- [x] **Lua**
- [x] **Makefile**
- [x] Manpage
- [x] **Markdown**
- [x] **MATLAB**
- [x] MemInfo (`/proc/meminfo`)
- [x] NAnt Build File
- [x] nginx
- [x] Nim
- [x] Nix
- [x] **Objective-C**
- [x] **Objective-C++**
- [x] **OCaml**
- [x] orgmode
- [x] **Pascal**
- [x] `/etc/passwd`
- [x] **Perl**
- [x] **PHP**
- [x] Plain Text
- [x] PowerShell
- [x] Protocol Buffer
- [x] Puppet
- [x] PureScript
- [x] **Python**
- [x] QML
- [x] **R**
- [x] Rego
- [x] **Regular Expression**
- [x] `requirements.txt`
- [x] resolv (`/etc/resolv.conf`)
- [x] **reStructuredText**
- [x] Robot Framework.
- [x] **Ruby**
- [x] Ruby Haml
- [x] **Ruby on Rails**
- [x] **Rust**
- [x] Salt State (SLS)
- [x] Sass
- [x] **Scala**
- [x] SCSS
- [x] SML
- [x] **SQL**
- [x] SSH Config
- [x] SSHD Config
- [x] Strace
- [x] Stylus
- [x] Swift
- [x] syslog
- [x] **Tcl**
- [x] Terraform
- [x] TeX
- [x] **Textile**
- [x] TOML
- [x] TypeScript
- [x] TypeScriptReact
- [x] varlink
- [x] Verilog
- [x] VimL
- [x] Vue Component
- [x] **XML**
- [x] **YAML**

It would be great if we could focus on the syntaxes in `sublimehq/Packages` first (marked in bold) as this would allow us to merge #1174 without having to worry (too much) about syntax highlighting regressions. Also, most of these are really popular languages, so it makes sense to have them in the test suite.

**Comment by adgai19** [comment]
Can I submit a python file for this issue?

**Comment by sharkdp** [comment]
> Can I submit a python file for this issue?

yes :+1:

**Comment by uduDudu** [comment]
I will submit a elixir file soon :)

**Comment by AkshatGadhwal** [comment]
i am not able to run `bash update.sh` it gives me error :
File "/usr/lib/python3.8/subprocess.py", line 1702, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'bat'

**Comment by dmbaturin** [comment]
@AkshatGadhwal @sharkdp I was also thinking about it. I think the script should first try to run a _locally_ built binary from `target/` if one is available. I'm not very familiar with Rust workflows yet though.

**Comment by CCreativeCND** [comment]
I'm also getting an error
```
Traceback (most recent call last):
  File "create_highlighted_versions.py", line 85, in <module>
    create_highlighted_versions(output_basepath=args.output)
  File "create_highlighted_versions.py", line 35, in create_highlighted_versions
    ["bat"] + BAT_OPTIONS + [source], stderr=subprocess.PIPE, env=env,
  File "/usr/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.6/subprocess.py", line 423, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'bat': 'bat'
```

**Comment by dmbaturin** [comment]
@CCreativeCND The solution is to add the `target/debug/` dir to your `$PATH`, or copy the binary `target/debug/bat` to somewhere in your `$PATH`. That's assuming you run `cargo build` wihout changing any build options (else the target may be something else than debug).

**Comment by Mithil467** [comment]
Hey @sharkdp :wave:
>We could probably improve the error message in the Python script.

This looks like a simple try catch statement. I'll be happy to submit a PR to add this if it helps :)

**Comment by AkshatGadhwal** [comment]
which files are supposed to be changed??
by running `bash update.sh` all the files in the 'highlights' folder are modified.

**Comment by sharkdp** [comment]
> which files are supposed to be changed??
> by running `bash update.sh` all the files in the 'highlights' folder are modified.
> 

Please make sure that you are running the latest version of bat. Only one new file should be generated.

**Comment by AkshatGadhwal** [comment]
> > which files are supposed to be changed??
> > by running `bash update.sh` all the files in the 'highlights' folder are modified.
> 
> Please make sure that you are running the latest version of bat. Only one new file should be generated.

i run `choco upgrade bat` but it is showing that you have the latest version  v0.16.0

**Comment by Mithil467** [comment]
@AkshatGadhwal What is the output for ``bat -V`?

**Comment by AkshatGadhwal** [comment]
@Mithil467 yes it is 0.16.0
![ss_bat_v](https://user-images.githubusercontent.com/58337809/95042987-25754d80-06f9-11eb-9c39-15fd30697dd2.PNG)

**Comment by AkshatGadhwal** [comment]
> @Mithil467 yes it is 0.16.0
> ![ss_bat_v](https://user-images.githubusercontent.com/58337809/95042987-25754d80-06f9-11eb-9c39-15fd30697dd2.PNG)

should I discard the changes in other files?

**Comment by Mithil467** [comment]
No, do not run chocho upgrade, but just `bat -V`. That should give you the version for the bat that is installe and detected on path.

**Comment by AkshatGadhwal** [comment]
> No, do not run chocho upgrade, but just `bat -V`. That should give you the version for the bat that is installe and detected on path.

I did that too.

![bat_V](https://user-images.githubusercontent.com/58337809/95043404-48543180-06fa-11eb-9042-81bc7d9afb27.PNG)

**Comment by sharkdp** [comment]
> I did that too.

Hm... I see that you opened a PR now anyway. Did it work for you? Or did you simply exclude all other files?

Maybe there is an incompatibility problem when running these scripts in Windows? Has someone else used Windows to make their change?

**Comment by Mithil467** [comment]
I pulled your branch locally and did ./update.sh. And I do see some changes in highlighted test.dart.
Diff shows -
![image](https://user-images.githubusercontent.com/41439841/95044651-94ed3c00-06fd-11eb-958f-b4a5e8fba0c3.png)

Also, the dart file has ^M characters (\r) in it, which might be an issue?

**Comment by AkshatGadhwal** [comment]
> > I did that too.
> 
> Hm... I see that you opened a PR now anyway. Did it work for you? Or did you simply exclude all other files?
> 
> Maybe there is an incompatibility problem when running these scripts in Windows? Has someone else used Windows to make their change?

I discarded other changes

**Comment by AkshatGadhwal** [comment]
> I pulled your branch locally and did ./update.sh. And I do see some changes in highlighted test.dart.
> Diff shows -
> ![image](https://user-images.githubusercontent.com/41439841/95044651-94ed3c00-06fd-11eb-958f-b4a5e8fba0c3.png)
> 
> Also, the dart file has ^M characters (\r) in it, which might be an issue?

yes, you are right \r is creating an issue.
to run `bash update.sh` I had to run `dos2unix update.sh`

**Comment by Mithil467** [comment]
Not the update.sh file, but your test.dart file.

**Comment by sharkdp** [comment]
Ok, this looks like a line ending problem indeed. I just opened https://github.com/sharkdp/bat/pull/1254 which might fix this (after it is merged and after you re-cloned the repository).

**Comment by AkshatGadhwal** [comment]
> Not the update.sh file, but your test.dart file.

so what I supposed to do know,
when I committed two files were created 
1. test.dart in highlighted/Dart folder 
2. test.dart in source/Dart folder

**Comment by Mithil467** [comment]
After #1254 is merged into master, you may pull the branch and update your fork's branch as well. Then rerun the changes.
Alternatively an easier way would be to run-
```
git remote add upstream http://github.com/sharkdp/bat
git fetch upstream
git checkout your_test_feature_branch
git reset upstream/master
```
This would update your local branch with the main repo's master.
Then you may re run the update.sh script and redo the process. You'll need to force push finally.

**Comment by sharkdp** [comment]
> After #1254 is merged into master

it has been merged by now.

**Comment by sharkdp** [comment]
It would be great if we could focus on the syntaxes in `sublimehq/Packages` first:
- [ ] ActionScript
- [ ] AppleScript
- [ ] ASP
- [ ] Batch File
- [x] C#
- [x] C++
- [ ] Clojure
- [ ] CSS
- [ ] D
- [ ] Diff
- [ ] Erlang
- [x] Go
- [ ] Graphviz
- [ ] Groovy
- [x] Haskell
- [ ] HTML
- [x] Java
- [x] JavaScript
- [x] JSON
- [x] LaTeX
- [ ] Lisp
- [ ] Lua
- [ ] Makefile
- [x] Markdown
- [ ] Matlab
- [ ] Objective-C
- [x] OCaml
- [ ] Pascal
- [ ] Perl
- [ ] PHP
- [x] Python
- [ ] R
- [ ] Rails
- [ ] Regular Expressions
- [ ] RestructuredText
- [x] Ruby
- [x] Rust
- [x] Scala
- [x] ShellScript
- [ ] SQL
- [ ] TCL
- [ ] Textile
- [x] XML
- [ ] YAML

This would allow us to merge #1174 without having to worry (too much) about syntax highlighting regressions. Also, most of these are really popular languages, so it makes sense to have them in the test suite.

**Comment by henil** [comment]
@sharkdp What do you think about taking example files from [learn x in y minutes](https://learnxinyminutes.com/)? We would still need to make them manually from their [repo](https://github.com/adambard/learnxinyminutes-docs) but it would be faster.

**Comment by sharkdp** [comment]
I'm not an expert on open source licenses, but it seems like the [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/deed.en_US) license is not a really great fit for us:

> ShareAlike — If you remix, transform, or build upon the material, you must distribute your contributions under the same license as the original.

It's not completely clear what this means for `bat` as a project if we were to modify the files.

**Comment by dmbaturin** [comment]
@sharkdp It's like GPL. You need to distribute modified versions of that file under the same license, CC-BY-SA. It only applies to that exact file in this case, since it's not a part of a larger work really and remains independent (by analogy, incorporating a CC-BY-SA song into a soundtrack of a CC-BY-NC-ND movie would be against the "share-alike" clause, but distributing songs under CC-BY-SA and other licenses on the same CD wouldn't).

**Comment by sharkdp** [comment]
@dmbaturin If that is the case, I'm also fine with adding files from this source. We should still check if they are really suited for our goal here. The large amount of comments in between is not too helpful, for example.

**Comment by dmbaturin** [comment]
@sharkdp Yeah, that's a much bigger obstacle. Examples from http://rosettacode.org (GFDL-licensed) can make much better placeholders until a comprehensive syntax test file can be found or made.

In fact, maybe a collection of _comprehensive_ syntax sample files should be a project of its own.

**Comment by sharkdp** [comment]
> In fact, maybe a collection of _comprehensive_ syntax sample files should be a project of its own.

That would be really helpful, yes :smile:

**Comment by henil** [comment]
@sharkdp removing comments shouldn't be that hard. May be I can strip all the comments out and make a new project as @dmbaturin mentioned. As [learn x in y minutes](https://learnxinyminutes.com/) seems to cover all syntax of a particular language.

Let me know if you find it interesting.

**Comment by sharkdp** [comment]
Ok, but maybe it should be discussed in a bit more detail before you spend too much time on this.

**Comment by henil** [comment]
Sure

**Comment by henil** [comment]
@sharkdp I managed to hack around [learnxinyminutes's repo](https://github.com/adambard/learnxinyminutes-docs) repo to extract needed part with my dirty python script. Here [syntax-samples](https://github.com/henil/syntax-samples).

I have also cleaned up the posts that where derivative of other languages. Or just explaination posts(like set theory...).

If we plan to go forward with it, It is a good starting point as it needs further processing. And I could use some help with that :)

**Comment by grg121** [comment]
Hi there! 

This project seems pretty good, I would like to contribute by adding a Dockerfile sample file. I will open a Pull Request as soon as possible.

**Comment by Saul-BT** [comment]
Can I submit a perl file for this issue?

**Comment by sharkdp** [comment]
> Can I submit a perl file for this issue?

That would be great! :+1:

**Comment by IsuruAriyarathne** [comment]
Can I submit a Java file for this issue?

**Comment by sharkdp** [comment]
Yes

**Comment by VictorAssunc** [comment]
Can I submit a Matlab file?

**Comment by zjr** [comment]
@sharkdp Do you have any thoughts about how you would add one of these for Regular Expressions? I tried writing a regular expression literal in JS and that gets highlighted well.

**Comment by sharkdp** [comment]
> @sharkdp Do you have any thoughts about how you would add one of these for Regular Expressions? I tried writing a regular expression literal in JS and that gets highlighted well.

I think it should work fine if you choose a filename with `.re` extension. The content would be just the plain regex.

**Comment by frankschmitt** [comment]
I'd like to submit a D file if that's ok.

**Comment by sharkdp** [comment]
Thanks to @henil and all of the other contributors, we have now syntax tests up and running for all main syntaxes (marked in bold) from [sublimehq/Packages](https://github.com/sublimehq/Packages/).

What does that mean?

We can now work towards updating sublimehq/Packages to the latest state with much more confidence regarding possible regressions. Unfortunately, `syntect`, our syntax highlighting engine, is not compatible with the latest format of the Sublime Text syntaxes, that is starting to be used in sublimehq/Packages. We will therefore likely have to maintain our own fork of that repo, that branches off from the currently used state in `bat`. We can then apply patches that are compatible with `syntect`/`bat` and have them checked automatically.

**Comment by Keats** [comment]
The `zola` users would be interested in helping with that fork I think as well. Being stuck on some old syntaxes means we do not get updates (eg `await` is not highlighted in the `st3` branch of the Rust syntax) or bugs (https://github.com/getzola/zola/issues/1241 which might be an issue in syntect actually, not 100% sure).

**Comment by sharkdp** [comment]
Thanks to @mohamed-abdelnour who added the remaining languages in #1668, this can be closed. Thank you to everyone involved!
